### PR TITLE
Create zipCodeToState Mapper for RADx

### PR DESCRIPTION
**Pull request by brick-green**
This PR adds a new mapper, ZipCodeToState, that utilizes the zip-code-data lookup table to add the patient state based on the zip code. Currently, the application will overwrite any incoming value based on the lookup results. To avoid overwriting a non-blank state value, this adds an additional field to the hl7-ingest-covid-19 schema, patient_state_zip_lookup. Additionally, this PR fixes a bug with the lookup mapper. The lookup mapper assumed that fields used in the mapper all used the same lookup table. That is an unsafe assumption. The PR fixes it so that the lookup columns are set in the parameters of the mapper removing the dependency.

Test Steps:
1. build and run all three test files through the application
     - `./prime data --input /PATH/TO/TEST/FILE.txt --input-schema HL7/hl7-ingest-covid-19-training  --output-dir /PATH/TO/OUTPUT/DIR/ --route`
2. Check output HL7
     - intrivo-test-missing-zip.txt: patient state in the output should be the same as input
     - intrivo-test-with-state.txt: patient state in the output should be same as input
     - intrivo-test-missing-state.txt: patient state in output should be different from input

Test Files
- [intrivo-test-missing-zip.txt](https://github.com/CDCgov/prime-reportstream/files/9237768/intrivo-test-missing-zip.txt)
- [intrivo-test-with-state.txt](https://github.com/CDCgov/prime-reportstream/files/9237769/intrivo-test-with-state.txt)
- [intrivo-test-missing-state.txt](https://github.com/CDCgov/prime-reportstream/files/9237770/intrivo-test-missing-state.txt)

## Changes
- added zipCodeToState mapper
- add column parameters for lookup mapper
- add unit test for zipCodeToState

### Testing
- [X] Tested locally?
- [X] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?

## Linked Issues
- Fixes #issue

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #5700

### CODE REVIEW

File: prime-router/src/main/kotlin/metadata/Mappers.kt [hunk 1]
```diff
@@ -468,16 +469,20 @@ class LookupMapper : Mapper {
...
-                values.forEach {
-                    val indexColumn = it.element.tableColumn
-                        ?: error("Schema Error: no tableColumn for element ${it.element.name}")
-                    tableFilter.equalsIgnoreCase(indexColumn, it.value)
+                values.forEachIndexed { index, elementAndValue ->
+                    // retrieve column name to use for lookup.
+                    // Specified in sender settings lookup mapper parameters
+                    val indexColumn = args[index.times(2).plus(1)].split(":")[1]
+                    if (indexColumn.isEmpty())
+                        "Schema Error: no tableColumn for element ${elementAndValue.element.name}"
```
Comment (TomNUSDS) [inline prime-router/src/main/kotlin/metadata/Mappers.kt:484]: Think this should be `error("Schema Error: no tableColumn for element ${elementAndValue.element.name}"`?

File: prime-router/src/main/kotlin/metadata/Mappers.kt [hunk 2]
```diff
@@ -932,6 +937,37 @@ class ZipCodeToCountyMapper : Mapper {
...
+/**
+ * [ZipCodeToStateMapper] runs a lookup using zip code and returns the single associated state. Null is returned if
+ * no zip code is provided, the zip code is not found, or if multiple records are found in the lookup.
+ */
+
+class ZipCodeToStateMapper : Mapper {
+    override val name = "zipCodeToState"
+
+    override fun valueNames(element: Element, args: List<String>): List<String> {
+        return args
```
Comment (TomNUSDS) [inline prime-router/src/main/kotlin/metadata/Mappers.kt:949]: My kotlin isn't super strong. What does this call do?

File: prime-router/src/test/kotlin/metadata/MapperTests.kt [hunk 3]
```diff
@@ -685,6 +685,60 @@ class MapperTests {
...
     }
 
+    @Test
+    fun `test zip code to state mapper`() {
+        val mapper = ZipCodeToStateMapper()
+        val csv = """
+            zipcode,state_abbr
+            32303,FL
+            42223,TN
+            42223,KY
```
Comment (TomNUSDS) [inline prime-router/src/test/kotlin/metadata/MapperTests.kt:695]: Maybe add some 5-4 zip codes (e.g. `32303-1234`) since code should handle them?

**Review by TomNUSDS — CHANGES_REQUESTED** [review]
Looks pretty great. I did have one thing that might be a bug with needing to return error().

File: prime-router/metadata/schemas/NM/nm-covid-19-csv.schema [hunk 4]
```diff
@@ -170,7 +170,7 @@ elements:
     type: TABLE
     table: LIVD-SARS-CoV-2
     tableColumn: Component
-    mapper: lookup(equipment_model_name)
+    mapper: lookup(equipment_model_name, $Column:Model)
```
[the remark this answers is not in the document]
Comment (brick-green) [reply prime-router/metadata/schemas/NM/nm-covid-19-csv.schema:173]: There was a bug in the lookup mapper that I found while working on the state mapper. This is part of the fix for that.

File: prime-router/metadata/schemas/HL7/hl7-ingest-covid-19-training.schema [hunk 5]
```diff
@@ -17,4 +17,4 @@ elements:
       - code: T
         display: T
       - code: T
-        display: P
+        display: P
```
[the remark this answers is not in the document]
Comment (brick-green) [reply prime-router/metadata/schemas/HL7/hl7-ingest-covid-19-training.schema:20]: I think this was incidental. I originally added a field to this file then decided it fit better elsewhere. Must have deleted the line feed that was there.

File: prime-router/src/main/kotlin/metadata/Mappers.kt [hunk 6]
```diff
@@ -468,16 +469,20 @@ class LookupMapper : Mapper {
...
                 val lookupTable = element.tableRef
                     ?: error("Schema Error: could not find table ${element.table}")
                 val tableFilter = lookupTable.FilterBuilder()
-                values.forEach {
-                    val indexColumn = it.element.tableColumn
-                        ?: error("Schema Error: no tableColumn for element ${it.element.name}")
-                    tableFilter.equalsIgnoreCase(indexColumn, it.value)
+                values.forEachIndexed { index, elementAndValue ->
+                    // retrieve column name to use for lookup.
+                    // Specified in sender settings lookup mapper parameters
```
[the remark this answers is not in the document]
Comment (brick-green) [reply prime-router/src/main/kotlin/metadata/Mappers.kt:481]: I like that idea. I'll add to the comment

File: prime-router/src/test/kotlin/metadata/MapperTests.kt (hunk 3 above)
Comment (brick-green) [reply prime-router/src/test/kotlin/metadata/MapperTests.kt:695]: good calll here too. Will do

File: prime-router/src/main/kotlin/metadata/Mappers.kt (hunk 2 above)
Comment (brick-green) [reply prime-router/src/main/kotlin/metadata/Mappers.kt:949]: That provides the list of fields used by the mapper. It has to be overwritten by each mapper. Some don't need to do any processing (like this one) but other's do(like the lookup mapper).

**Review by TomNUSDS — APPROVED** [review]
Approved. Thanks for the fixing this state matching issue!

Anything that needs to be communicated to others before it goes live?

**Pull request merged**